conf.cpp

```diff
@@ -0,0 +1,276 @@
+/*
+ * This file is part of the MAVLink Router project
+ *
+ * Copyright (C) 2017  Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "conf.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "log.h"
+#include "util.h"
+
+struct config {
+    char *key;
+    char *value;
+    struct config *next;
+};
+
+struct section {
+    char *name;
+    struct config *configs;
+    struct config *current_config;
+    struct section *next;
+};
+
+ConfFile::~ConfFile()
+{
+    struct section *section;
+    struct config *config;
+
+    section = _sections;
+    while (section) {
+        struct section *tmp = section;
+
+        config = section->configs;
+        while (config) {
+            struct config *tmp_config = config;
+            free(config->key);
+            // config->value is on the same allocation as config->key, no need to free it
+            config = config->next;
+            free(tmp_config);
+        }
+
+        free(section->name);
+        section = section->next;
+        free(tmp);
+    }
+}
+
+int ConfFile::parse_file()
+{
+    FILE *file;
+    char *entry = NULL;
+    int line = 0, ret = 0;
+    size_t n = 0;
+    ssize_t read;
+
+    assert(_filename);
+
+    file = fopen(_filename, "re");
+    if (!file) {
+        log_error_errno(errno, "Could not open conf file '%s' (%m)", _filename);
+        return -EIO;
+    }
+
+    while ((read = getline(&entry, &n, file)) >= 0) {
+        line++;
+
+        _trim(&entry);
+
+        switch (entry[0]) {
+        case ';':
+        case '#':
```

Comment:
people can just comment on the line above.   commenting inline in the end of the line just causes trouble... you would need to add a way to escape # when it's needed